example/bwmf/bwmf_task.go

```diff
@@ -74,12 +74,150 @@ type bwmfTask struct {
 	childrenReady map[uint64]bool
 
 	dShard, tShard Shard
-	d, t           []*bwmfData
+	d, t           Shard
+
+	// parameters for projected gradient methods
+	sigma, alpha, beta    float64
+	maxIterInSubProblem   int
+	maxIterInFindingAlpha int
 }
 
 // These two function carry out actual optimization.
-func (t *bwmfTask) updateDShard() {}
-func (t *bwmfTask) updateTShard() {}
+func (t *bwmfTask) updateDShard() {
+	var newDShard Shard = t.dShard
+	var ok bool
+	for iter := 0; iter < t.maxIterInSubProblem; iter++ {
+		newDShard, ok = t.solveSubproblem(&t.rowShard, &newDShard, &t.t)
+		// TODO(baigang): dealing with ok==false
+		// TODO(baigang): logging
+	}
+	t.dShard = newDShard
+}
+func (t *bwmfTask) updateTShard() {
+	var newTShard Shard = t.tShard
+	var ok bool
+	for iter := 0; iter < t.maxIterInSubProblem; iter++ {
+		newTShard, ok = t.solveSubproblem(&t.columnShard, &t.tShard, &t.d)
+		// TODO(baigang): dealing with ok==false
+		// TODO(baigang): logging
+	}
+	t.tShard = newTShard
+}
+
+// calculates $matrix^T * matrix$
+func matrixSelfTransMult(matrix *Shard) Shard {
+	var result Shard
+	// TODO(baigang): implementation
+	// length := len(matrix[0])
+	return result
+}
+
+// calculates $a * b$ where $b$ is a sparseVec
+func matrixSparseMult(a *Shard, b *[]sparseVec) Shard {
```

Comment:
Yes, this should be reusable. Is it ok to put these operations into example/bwmf/matrix.go?